contracts/eosio.system/include/eosio.system/eosio.system.hpp

```diff
@@ -282,6 +291,18 @@ namespace eosiosystem {
       EOSLIB_SERIALIZE( producer_info2, (owner)(votepay_share)(last_votepay_share_update) )
    };
 
+   // Defines finalizer_key info structure to be stored in finalizer_keys info table, added after version 6.0
+   struct [[eosio::table, eosio::contract("eosio.system")]] finalizer_key_info {
+      name                            producer;
+      std::string                     active_finalizer_key; // the currently active key
+      std::unordered_set<std::string> registered_finalizer_keys; // including the active_finalizer_key
+
+      uint64_t primary_key()const { return producer.value; }
```

Comment:
We should not put all the finalizer keys of a producer in the same table record.

Additionally, we need to check for duplicate keys across all producers, not just for each producer.

I think this means we need a table to capture all finalizer keys that has a secondary index which is based on a hash of the public key data (in serialized form). That allows for a quick global check to see if the key is already registered. The primary key can be some auto-incrementing ID number.

We may also need another secondary index that sorts based on the producer name. This shouldn't be necessary for the contract. But it would be useful for users to explore the keys they have currently registered in a block explorer to know which one to delete or activate. Along those lines, it is useful to also store the base64url encoded version of the public key in that record. In fact, the binary version may not even be necessary to store.

We would still have a separate table that has one record per finalizer. This would at a minimum store the ID of the finalizer key that is currently active for the finalizer and store a count of the number of finalizer keys registered for the finalizer. The count is necessary to prevent deleting an active finalizer key unless it was the last registerd finalizer key of the finalizer. It would would also make sense to cache the active finalizer key in that data structure to avoid needing another lookup of a table record of the key when constructing a finalizer policy; especially if the table record of the key only stores the base64url encoding.